core/src/main/java/org/apache/iceberg/jdbc/JdbcUtil.java

```diff
@@ -99,4 +136,40 @@ public static Properties filterAndRemovePrefix(Map<String, String> properties,
 
     return result;
   }
+
+  public static String updatePropertiesStatement(Map<String, String> properties) {
```

Comment:
I'm impressed that you came up with a way to do this in a single statement. I didn't think about this when I made my original suggestion because I was thinking about using INSERT as an UPSERT, rather than breaking the action into two stages. I think the actual SQL works well, but it is really difficult to see how the code constructs it, especially the use of `{}` and replace.

Honestly, I think that most of the parameterizing SQL done in this class isn't worth it. The SQL is really hard to read because of all the string manipulation. I'd probably prefer to just inline most of it so at least this method is readable. Like this:

```java
sqlStatement.append("UPDATE ... SET value = CASE");
for (int i = 0; i < size; i += 1) {
  sqlStatement.append(" WHEN key = ? THEN ?")
}
sqlStatement.append(" END");
...

Reply:
Yeah, it took some effort to get it working right. I like the approach you mentioned. I'll clean it up to help simplify.